gobblin-api/src/main/java/gobblin/publisher/DataPublisher.java

```diff
@@ -54,8 +82,7 @@ public abstract void publishMetadata(Collection<? extends WorkUnitState> tasks)
    * @param states is a {@link Collection} of {@link WorkUnitState}s.
    * @throws IOException if there is a problem with publishing the metadata or the data.
    */
-  public void publish(Collection<? extends WorkUnitState> states)
-      throws IOException {
+  public void publish(Collection<? extends WorkUnitState> states) throws IOException {
     publishMetadata(states);
```

Comment:
This may lead to data loss. If the metadata includes checkpoints and the publish of data fails, on the next run will use the new checkpoints.

We should publish data first. Generally, duplicate data is better than data loss.